core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsStatementCustomizer.java

```diff
@@ -13,20 +13,48 @@
  */
 package org.jdbi.v3.core.statement;
 
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
 import java.sql.PreparedStatement;
-import java.sql.SQLException;
 import java.util.Set;
 
-import org.jdbi.v3.core.argument.NullArgument;
+import org.jdbi.v3.core.argument.Argument;
+import org.jdbi.v3.core.internal.exceptions.Unchecked;
 
 class DefineNamedBindingsStatementCustomizer implements StatementCustomizer {
     @Override
-    public void beforeTemplating(PreparedStatement stmt, StatementContext ctx) throws SQLException {
+    public void beforeTemplating(PreparedStatement stmt, StatementContext ctx) {
         final Set<String> alreadyDefined = ctx.getAttributes().keySet();
         final Binding binding = ctx.getBinding();
+        final SetNullHandler handler = new SetNullHandler();
         binding.getNames().stream()
             .filter(name -> !alreadyDefined.contains(name))
             .forEach(name -> binding.findForName(name, ctx).ifPresent(
-                    a -> ctx.define(name, a instanceof NullArgument ? false : true)));
+                    a -> handler.define(name, a, ctx)));
+    }
+
+    private static class SetNullHandler implements InvocationHandler {
+        private final PreparedStatement fakeStmt = (PreparedStatement)
+                Proxy.newProxyInstance(Thread.currentThread().getContextClassLoader(), new Class<?>[] {PreparedStatement.class}, this);
+        private boolean setNull;
+        private boolean setCalled;
+
+        @Override
+        public Object invoke(Object proxy, Method method, Object[] args) {
+            setCalled = method.getName().startsWith("set");
```

Comment:
Might not be a single method call. Should probably change this to `||=`

Reply:
Something like this, rather:

```java
if (method.getName().startsWith("set")) {
  setCalled = true;
  ...etc
}
```

Reply:
I'd personally go for `|=` to save the `if` block but the logic change makes sense.

Reply:
I think it needs to be in an if-block, otherwise the test for null on the following lines isn't valid.

Reply:
Pushed fix